shell/services/node/node_service.cc

```diff
@@ -0,0 +1,152 @@
+    : node_bindings_(
+          NodeBindings::Create(NodeBindings::BrowserEnvironment::kUtility)),
+      electron_bindings_(
+          std::make_unique<ElectronBindings>(node_bindings_->uv_loop())) {
+  if (receiver.is_valid())
+    receiver_.Bind(std::move(receiver));
+  if (auto* thread = content::UtilityThread::Get()) {
+    thread->BindHostReceiver(
+        node_service_host_remote_.BindNewPipeAndPassReceiver());
+  }
+}
+
+NodeService::~NodeService() {
+  // Destroy node platform.
+  node_env_->env()->set_trace_sync_io(false);
+  js_env_->OnMessageLoopDestroying();
+  node::Stop(node_env_->env());
+}
+
+void NodeService::Initialize(node::mojom::NodeServiceParamsPtr params) {
+  if (initialized_)
+    return;
+
+  initialized_ = true;
+
+  if (!params->environment.empty()) {
+    std::unique_ptr<base::Environment> env(base::Environment::Create());
+    for (const auto& variable : params->environment) {
+      env->SetVar(variable->name, variable->value);
+    }
+  }
+
+  js_env_ = std::make_unique<JavascriptEnvironment>(node_bindings_->uv_loop());
+
+  v8::HandleScope scope(js_env_->isolate());
+
+  node_bindings_->Initialize();
+
+  // Append program path for process.argv0
+  auto program = base::CommandLine::ForCurrentProcess()->GetProgram();
+#if defined(OS_WIN)
+  params->args.insert(params->args.begin(), base::WideToUTF8(program.value()));
+#else
+  params->args.insert(params->args.begin(), program.value());
+#endif
+
+  // Create the global environment.
+  node::Environment* env = node_bindings_->CreateEnvironment(
+      js_env_->context(), js_env_->platform(), params->args, params->exec_args);
+  node_env_ = std::make_unique<NodeEnvironment>(env);
+
+  env->set_trace_sync_io(env->options()->trace_sync_io);
+
+  // Add Electron extended APIs.
+  electron_bindings_->BindTo(env->isolate(), env->process_object());
+
+  // Add entry script to process object.
+  gin_helper::Dictionary process(env->isolate(), env->process_object());
+  process.SetReadOnly("_serviceStartupScript", params->script);
+  process.SetHidden("_postMessage",
+                    base::BindRepeating(&NodeService::PostMessage,
+                                        weak_factory_.GetWeakPtr()));
+
+  // Load everything.
+  node_bindings_->LoadEnvironment(env);
+
+  // Wrap the uv loop with global env.
+  node_bindings_->set_uv_env(env);
+
+  // Setup microtask runner.
+  js_env_->OnMessageLoopCreated();
+
+  // Run entry script.
+  node_bindings_->PrepareEmbedThread();
+  node_bindings_->StartPolling();
+}
+
+void NodeService::PostMessage(v8::Local<v8::Value> message_value) {
+  if (!node_service_host_remote_)
+    return;
+
+  v8::Isolate* isolate = js_env_->isolate();
+  blink::TransferableMessage transferable_message;
+  electron::SerializeV8Value(isolate, message_value, &transferable_message);
+  node_service_host_remote_->ReceivePostMessage(
+      base::Process::Current().Pid(), std::move(transferable_message));
```

Comment:
We should not allow the child process to specify its own pid; the identity of the message sender should be tied to the pipe in the main process.

Reply:
https://github.com/electron/electron/pull/34980/commits/e3b4c229c389d8047dffdc168baa7439912e92bc